README.md
# sanic-healthchecks
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![PyPI version](https://badge.fury.io/py/sanic-healthchecks.svg)](https://badge.fury.io/py/sanic-healthchecks)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/sanic-healthchecks.svg)](https://pypi.python.org/pypi/sanic-healthchecks/)
![PyPI - Downloads](https://img.shields.io/pypi/dm/sanic-healthchecks.svg)

sanic-healthchecks makes it easy for you to start a healthcheck server on a
different thread than your actual Sanic application.

## Installation

`pip3 install sanic-healthchecks`

## Healthcheck Example
```python
from sanic import Sanic
from sanic.response import json

from sanic_healthchecks import start_healthcheck_server, healthcheck_response

APP = Sanic()


async def healthcheck_handler(_):
    data = {"status": "ok"}
    return healthcheck_response(data)


@APP.route("/")
async def root(_):
    return json({"example_of": "a very simple healthcheck"})


if __name__ == "__main__":
    start_healthcheck_server(healthcheck_handler)
    APP.run(host="0.0.0.0", port=8000)
```

Your Sanic application will now respond on to healthchecks on a different port:
```
⇒  curl http://localhost:8000 -i
HTTP/1.1 200 OK
Connection: keep-alive
Keep-Alive: 5
Content-Length: 17
Content-Type: application/json

{"hello":"world"}

⇒  curl http://localhost:8082 -i
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Content-Length: 16
Date: Sun, 03 Mar 2019 20:55:52 GMT
Server: Python/3.7 aiohttp/3.5.4

{"status": "ok"}
```

## Changelog
[Release Changelogs.](https://github.com/abatilo/sanic-healthchecks/blob/master/CHANGELOG.md)

## License
[Apache 2.0](https://github.com/abatilo/sanic-healthchecks/blob/master/LICENSE)

## But why?
Why would you want to run your healthchecks on a different thread, as opposed to creating another endpoint on your actual Sanic server?

Great question, internet stranger, and I have a few answers.

By running your healthchecks separately, we maintain a strong separation of
concerns. Since Sanic runs on a single thread, then any time you need to
respond to healthchecks, you're actually taking compute time away from the
event loop that is powering the actual requests that your application is there
to serve. Likewise, the state of your actual application is not going to affect
the healthchecks. There's a few camps of thought on this subject.

Some people say that if your web service isn't capable of responding to your
healthcheck probe, then the service shouldn't be considered healthy. I can
totally understand and respect this perspective, and if this is how you feel,
then there's no need to use sanic-healthchecks.

On the other hand, if you're like me, you've convinced yourself that the point
of healthchecks isn't purely to determine if the service can respond, but also
to determine if your service has everything that it needs from downstream
dependencies. If your requests are taking so long that the readiness or
liveness probes are timing out, that could mean that your service is unhealthy,
but it also could be a symptom of services that have long running requests.

Since we can run the healthchecks on a different web server entirely, we have
the ability to check that all of the downstream dependencies, like databases
and other services, are available. This helps narrow the problems with why a
service might be in a degraded state.

I would even make an argument that an increase in response latency could be a
metric that you use for automatically scaling your service. Treating it as a
way to kill instances makes it much fuzzier in terms of how to interpret the
increase in latency.

Another great reason to run your healthchecks on a different server is so that
you can assign a different port to this new server. This is valuable because
your healthchecks might actually have debug information in them that should not
be exposed to the same groups of people who are able to consume the main
service. By putting healthchecks on a different port, you can make sure to map
your load balancer to **not** include this healthcheck port.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

## [1.0.0] - 2019-05-29
- Nothing changed. This library has been used in production long enough to feel
  confident in it now.

## [0.2.0] - 2019-03-09
### Changed
- Renamed deploy task in `.circleci/config.yml` to publish
### Added
- Added `--no-dev` in `.circleci/config.yml` deploy task
- Expose wrapper for aiohttp web response, instead of leaking that we're using aiohttp under the hood
- Badge for downloads per month added to README.md
- Example for multiple healthchecks running concurrently
### Fixed
- CHANGELOG.md had 0.1.0 listed twice

## [0.1.1] - 2019-03-03
### Changed
- Went from MIT to Apache 2.0 License, to match aiohttp
- Example now uses `healthcheck_handler` instead of just `handler`
### Added
- .github directory for basic templates
- Keywords and classifiers for PyPI listing
- LICENSE file
- README.md file

## [0.1.0] - 2019-03-03
### Added
- Initial version. Includes sanic_healthchecks.start_healthchecks_server


sanic_healthchecks/core.py
"""The core of running a healthcheck server on a different thread.

It's useful to run the healthchecks on a different web server entirely, because
it means that we're not going to take away any resources from the thread that
is running the actual Sanic application itself.

The healthchecks should report that the downstream and required dependencies of
a service are still accessible from the host or container that is running the
actual Sanic application.
"""
import asyncio
import json
import threading
from aiohttp import web


# Disabled because pylint incorrectly parses multiline arguments like this:
# https://github.com/PyCQA/pylint/issues/289
# pylint: disable=C0330
def healthcheck_response(
    data,
    *,
    text=None,
    body=None,
    status=200,
    reason=None,
    headers=None,
    content_type="application/json",
    dumps=json.dumps,
):
    """A response for a healthcheck."""
    return web.json_response(
        data,
        text=text,
        body=body,
        status=status,
        reason=reason,
        headers=headers,
        content_type=content_type,
        dumps=dumps,
    )


async def main(handler, host, port):
    """The main entrypoint for running an aiohttp server.

    This function will start an async compatible aiohttp server. There's no
    other fancy logic or configuration. The handler that is passed in will be
    available as an endpoint at the root path of whatever port gets specified.

    Requests will be made by sending an HTTP GET verb to:
    http://{host}:{port}/
    """
    server = web.Server(handler)
    runner = web.ServerRunner(server)
    await runner.setup()
    site = web.TCPSite(runner, host, port)
    await site.start()

    print(f"======= Listening for healthchecks on http://{host}:{port}/ ======")

    # pause here for very long time by serving HTTP requests and
    # waiting for keyboard interruption
    while True:
        await asyncio.sleep(100 * 3600)


def start_healthcheck_server(handler, host="0.0.0.0", port=8082):
    """Start the actual web server to respond to healthcheck requests.

    We need to make sure that this web server is non-blocking, so that the
    Sanic application can be instantiated as expected.
    """
    loop = asyncio.get_event_loop()
    httpd = threading.Thread(
        target=lambda: loop.run_until_complete(main(handler, host, port))
    )
    httpd.daemon = True
    httpd.start()


examples/multi.py
"""Basic sample for adding a healthcheck server to a Sanic application.
"""
import asyncio
from http import HTTPStatus
from sanic import Sanic
from sanic.response import json

from sanic_healthchecks import start_healthcheck_server, healthcheck_response

APP = Sanic()


async def healthcheck_task1():
    """A healthcheck representing a database check."""
    await asyncio.sleep(0.1)
    return ("database is good!", HTTPStatus.OK)


async def healthcheck_task2():
    """A healthcheck representing trying to send a request to another server."""
    await asyncio.sleep(0.15)
    return ("required service is good!", HTTPStatus.OK)


async def healthcheck_task3():
    """A healthcheck representing trying to verify valid AWS access."""
    await asyncio.sleep(0.5)
    return ("AWS credentials are good!", HTTPStatus.OK)


async def healthcheck_handler(_):
    """Demonstrates a simple healthcheck example.
    """
    tasks = (healthcheck_task1(), healthcheck_task2(), healthcheck_task3())
    responses = [
        {"message": msg, "status": status}
        for (msg, status) in await asyncio.gather(*tasks)
    ]
    overall_status = (
        200 if all(resp["status"].value == 200 for resp in responses) else 500
    )
    overall = {"status": overall_status, "responses": responses}
    return healthcheck_response(overall, status=overall_status)


@APP.route("/")
async def root(_):
    """Boilerplate Sanic endpoint.
    """
    return json({"example_of": "multiple healthchecks running concurrently"})


if __name__ == "__main__":
    start_healthcheck_server(healthcheck_handler)
    APP.run(host="0.0.0.0", port=8000)
